### ISSUE TITLE
MD060: style "compact" requires one space per side for content cells but one space total for empty cells

### ISSUE THREAD

**Issue by Spenhouet** [DavidAnson/markdownlint#2196]
## Summary

`MD060` style `compact` applies a different padding rule to empty cells than to cells with content. A cell with content needs one space on each side, while an empty cell is only allowed one space in total. `doc/md060.md` describes the style as "avoids extra padding with a single space around cell content", which reads as one space per side, so the empty-cell behaviour looks unintended.

## Reproduction

markdownlint 0.41.1, config `{"default": false, "MD060": {"style": "compact"}}`.

Cells with content, one space per side is required:

| cell | violations |
| :--- | :--- |
| `\|a\|` | 2, missing space to the left and right |
| `\| a \|` | 0 |
| `\|  a  \|` | 2, extra space to the left and right |

Empty cells, only one space in total is accepted:

| cell | violations |
| :--- | :--- |
| `\| \|` | 0 |
| `\|  \|` | 2, extra space to the left and right |
| `\|   \|` | 2 |

So `| a |` passes but `|  |` does not, although both have exactly one space between the pipe and the (empty) content on each side.

Full document that reproduces it:

```markdown
# T

|  |  |
| :--- | :--- |
| a | b |
```

```
line 3: MD060/table-column-style [Table pipe has extra space to the right for style "compact"]
line 3: MD060/table-column-style [Table pipe has extra space to the left for style "compact"]
(twice, once per empty cell)
```

`--fix` rewrites line 3 to `| | |`.

## Cause

In `lib/md060.mjs` the check walks `tableCellDivider`, `tableContent` and `whitespace` tokens and, for each divider, compares the adjacent `whitespace` token against a length of 1:

```js
if (previous.type === "whitespace") {
  if (previous.text.length !== 1) { /* extra space to the left */ }
```

For a cell with content there are two separate whitespace tokens, one on each side of the `tableContent` token, and each is checked once against 1.

For an empty cell there is no `tableContent` token, so micromark emits a single `whitespace` token spanning the whole gap. That one token is then checked twice, once as the right padding of the preceding divider and once as the left padding of the following divider, each time against a length of 1. A gap of two spaces, which is one space per side, therefore fails both checks, and a gap of three spaces produces the same two errors.

## Why it matters

Table formatters commonly implement padding as "one space, content, one space" applied uniformly to every cell, which yields two spaces for an empty cell. Output of that shape is internally consistent but MD060 reports it under every style, including `any`, and the fix rewrites it. Where such a formatter also owns the files, the two tools revert each other on every save.

In my case that is roughly 59,600 reported violations across 1,818 of 2,142 files, all from empty cells alone, in a Markdown documentation repository whose tables are formatted by an editor plugin. The only self-consistent alternative is to switch the formatter to no padding at all and use style `tight`, which costs a lot of readability for documents that are read as plain text and in diffs.

## Suggestion

Accept a two-space gap in an empty cell under `compact`, on the basis that it is one space per side and matches the documented wording. Continuing to accept the current one-space form would keep existing files valid.

If you would rather not change `compact`, an option on the rule or an additional style would also solve it.

I am happy to open a pull request for whichever shape you prefer, including doc and test updates. I did not send one straight away because this changes rule semantics and that seemed like your call to make first.

**Comment by DavidAnson** [comment]
The current behavior is deliberate. The style is called "compact" and attempts to maintain a single space between cell dividers and cell content - or the nearest divider when no content is present. Analogously, the style "tight" maintains no space in the same scenarios. The extra second space you propose serves no purpose other than making things LESS compact. What's more, all of your empty cell examples above are rendered by GitHub Mobile as a single space which seems in keeping with the same spirit.

What I'd be curious to know is why whatever tool you are using thinks 2 spaces is appropriate here. If it's just a consequence of a simple implementation as you suggest in "Why it matters", that does not feel compelling to me.

**Comment by Spenhouet** [comment]
We are using our markdown documents via VS Code + Obsidian. Hence the separate formatters. And I'm trying to make both produce a matching format (not disagree with each other).

I also opened a respective PR on the other formatter to allow it to produce the same formatting as markdownlint is producing: https://github.com/fangface-hub/obsidian_table_formatter/pull/11

For us for now we need to disable the MD060 rule as working on the documentation is primarily done via Obsidian. But it would be great if markdownlint would also support the same format.

**Comment by DavidAnson** [comment]
It looks like the Obsidian formatter is going to make the change you propose to emit a single space for empty table cells, so I will hold off on doing anything with this issue as it should no longer be relevant.
